src/endpoints/patch-session.js

```diff
@@ -14,29 +15,160 @@ module.exports = cors(async (req, res) => {
+    if (patches.length) {
+        await asyncForEach(patches, async (patch) => {
+            const pathArray = patch.path.split('/')
+            const indexToRemove = pathArray[2]
+
+            await db.prepare('DELETE FROM sample_state WHERE session_short_id = ? AND session_sample_id = ? ').run(sessionId, indexToRemove);
+            await db.prepare(`
+                INSERT INTO sample_state 
+                    (session_short_id, session_sample_id, version, content, summary, annotation) 
+                SELECT 
+                    session_short_id, session_sample_id - 1, version + 1, content, summary, annotation
+                FROM 
+                    latest_sample_state
+                WHERE 
+                    session_short_id = ? AND session_sample_id > ?`).run(sessionId, indexToRemove);
+
+            await db.prepare(`
+                DELETE FROM sample_state 
+                WHERE sample_state_id IN (
+                    SELECT sample_state_id
+                    FROM sample_state
+                    WHERE session_sample_id = (
+                        SELECT session_sample_id
+                        FROM latest_sample_state 
+                        WHERE session_short_id = ? 
+                        ORDER BY session_sample_id DESC 
+                        limit 1
+                    )
+                )`).run(sessionId);
+        })
+    }
+})
+
+const patchSamplesAnnotation = (async (sessionId, samplePatches) => {
+
+    const sampleIds = {}
+    samplePatches.forEach(patch => {
+        const pathArray = patch.path.split('/')
+        const sessionSampleId = pathArray[2]
+        sampleIds[sessionSampleId] = {sessionSampleId}
+    })
+
+    const samplesArray = []
+    const samples = db.prepare('SELECT * FROM latest_sample_state WHERE session_short_id = ? ORDER BY session_sample_id ASC').all(sessionId);
+    samples.forEach(sample => {
+        const content = JSON.parse(sample.content)
+        const annotation = JSON.parse(sample.annotation)
+        samplesArray.push({
+            ...content,
+            annotation: annotation,
+            version: parseInt(sample.version),
+            session_sample_id: sample.session_sample_id
+        })
+    })
+    let newJSON = {samples: samplesArray}
+    fjp.applyPatch(newJSON, samplePatches)
+
+    const queries = []
+    Object.keys(sampleIds).forEach(sampleId => {
+        const sample = samplesArray[sampleId]
+        const session_sample_id = sample.session_sample_id
+        const version = sample.version + 1
+        const annotation = sample.annotation
+
+        delete sample.session_sample_id
+        delete sample.version
+        delete sample.annotation
+
+        queries.push(
+            db.prepare('INSERT INTO sample_state (session_short_id, session_sample_id, version, annotation, content) VALUES (?, ?, ?, ?, ?)')
+                .run(sessionId, session_sample_id, version, JSON.stringify(annotation), JSON.stringify(sample)));
```

Comment:
you should do an update on the sample, not an insert

Reply:
yup